### ISSUE TITLE
API/DEPR: DataFrame/Series.replace is too complex.

### ISSUE THREAD

**Issue by simonjayhawkins** [pandas-dev/pandas#33302]
in https://github.com/pandas-dev/pandas/issues/17673#issue-260396559 @jorisvandenbossche wrote

>Underlying reason is that this function of course can do way too many things at the same time (or the same things in too many different ways) ... (orthogonal to this, we could maybe also think if certain functionality could be moved into its own function).

opened this issue to discuss.

**Comment by simonjayhawkins** [comment]
the current behaviour when `value` is `None` is surprising from an end-user point of view #19998. 

https://github.com/pandas-dev/pandas/issues/19998#issuecomment-372030943 @reidy-p wrote

>These are good suggestions. I think it would be good if the method parameter is only used when explicitly called rather than by default because it's quite easy to get unexpected behaviour at present as you have shown. If it doesn't break the API substantially I would prefer if the behaviour of s.replace('a', None) is the same as s.replace({'a': None}).

**Comment by simonjayhawkins** [comment]
could the `regex` parameter be deprecated and use str.replace instead? DataFrames don't have a str accessor, but this is a user request see #31850

**Comment by jbrockmendel** [comment]
+1 for simplifying

**Comment by jbrockmendel** [comment]
After #33279 the follow-up removes the `filter` kwarg from `Block.replace`, which lets us simplify the implementatin of NDFrame.replace.  Doesnt affect the API, but pertinent.

**Comment by jbrockmendel** [comment]
I'm leaning towards deprecating the ffill/bfill behavior with value=None (in part bc it has hidden bugs bc Series._replace_single uses .values instead of ._values).

In my first attempt at deprecating I thought we could tell the user to do `obj.replace(to_replace, np.nan).ffill()`, but that actually messes up in cases where obj already has NAs that we dont want to fill.

**Comment by jbrockmendel** [comment]
I propose the following, which we can do any subset of:

1) deprecate the "method" keyword in replace
2) deprecate the value=lib.no_default case (possibly excepting dict-like to_replace)
3) deprecate the regex keyword and introduce a method for that.  Or maybe there is something users can do with a string accessor to avoid the need for a new method?

**Comment by jbrockmendel** [comment]
method and limit keywords are gone.  @simonjayhawkins do you think there is more that we can reasonably simplify here?  Also @jorisvandenbossche since you wrote the original comment linked in the OP.

**Comment by jbrockmendel** [comment]
Closing as complete

**Comment by jorisvandenbossche** [comment]
Quickly checking the current docstring, we have this section for `to_replace`:

```
Parameters
----------
to_replace : str, regex, list, dict, Series, int, float, or None
    How to find the values that will be replaced.

    ...

    * None:

        - This means that the `regex` argument must be a string,
          compiled regular expression, or list, dict, ndarray or
          Series of such elements. If `value` is also ``None`` then
          this **must** be a nested dictionary or Series.
```

Is that something that is needed? Does that enable anything else than just passing that value to `to_replace` with `regex=True`? 
(the `regex` parameter's type annotation also says `bool`, which seems wrong)

(but if this is worth checking, can also be a separate issue, I have never used replace with a regex myself, so no clear opinion)
